### PR TITLE
PEDS-1379 - standardize the payload

### PR DESCRIPTION
**Pull request by kellcuc**
updating the way of standardize the payload. You can follow the testing steps in readme.md file and run the test that was in place.

### CODE REVIEW

File: pcdcutils/gen3/__init__.py [hunk 1]
```diff
@@ -36,25 +36,38 @@ def get_gen3_service_header(self):
...
         sm = SignatureManager(key=private_key)
-        return sm.sign(payload)
-            
+
+        if isinstance(payload, str):
+            # Legacy/test mode
+            standardized_payload = payload
+        else:
+            # Standardized request
+            method = payload.method
```
Comment (paulmurdoch19) [inline pcdcutils/gen3/__init__.py:61]: what object is payload supposed to be? is it from the request library? We should generate a test for this

**Review by paulmurdoch19 — CHANGES_REQUESTED** [review]
Other than creating a test for the new data format this looks good

File: pcdcutils/gen3/__init__.py (hunk 1 above)
Comment (kellcuc) [reply pcdcutils/gen3/__init__.py:61]: make_gen3_signature takes either a raw string (in tests) or a full request object (in production). The current tests use string input like body = 'aaaaa', which works for older compatibility. 

I’ll add a test for the prod request object to make that use case explicit and verified.

Comment (kellcuc) [reply pcdcutils/gen3/__init__.py:61]: We should have a general consistent way to turn on logging across the board, maybe a future item :)
Screenshot 2025-04-21 at 6 49 21 PM
Screenshot 2025-04-21 at 6 52 30 PM

File: pcdcutils/gen3/__init__.py [hunk 2]
```diff
@@ -36,25 +36,37 @@ def get_gen3_service_header(self):
...
+            path = payload.path
+            body = payload.get_data(as_text=True) if method in ['POST', 'PUT', 'PATCH'] else ''
```
Comment (paulmurdoch19) [inline pcdcutils/gen3/__init__.py:63]: actually the use of get_data might be a problem because its a flask specific method and gearbox-middleware and gearbox-matching are running fastapi servers

File: pcdcutils/gen3/__init__.py [hunk 3]
```diff
@@ -36,25 +36,37 @@ def get_gen3_service_header(self):
...
+            standardized_payload = f"{method} {path}\nGen3-Service: {service_name}"
+            if body:
+                standardized_payload += f"\n{body}"
+
+        return sm.sign(standardized_payload)
+
 
     def valid_gen3_signature(self, payload='', config=None):
```
Comment (paulmurdoch19) [inline pcdcutils/gen3/__init__.py:71]: Also it seems we should make changes to this method as well to reverse the process of the make_gen3_signature and construct what the standardized_payload should be then verify the signature with that

**Review by paulmurdoch19 — CHANGES_REQUESTED** [review]

File: pcdcutils/gen3/__init__.py (hunk 2 above)
Comment (kellcuc) [reply pcdcutils/gen3/__init__.py]: Adjustment made thank you for that catch. Added some notes to possible future development as well.

File: pcdcutils/gen3/__init__.py (hunk 3 above)
Comment (kellcuc) [reply pcdcutils/gen3/__init__.py]: Great call, thanks! I have added a shared method now called build_standardized_payload

**Comment by paulmurdoch19** [comment]
@kellcuc Were you able to install this branch into (amanuensis and fence) and (gearbox-middleware and gearbox-matching) and test the make_gen3_signature and valid_gen3_signature methods to make sure this works as expected for both Flask and Fastapi? I would test it using helm as well. If so then this looks good to me

**Comment by grugna** [comment]
Just following up on this. Has it been tested?

**Comment by grugna** [comment]
Did you check if the downgrade of gen3 impacted other methods in this repo?

**Comment by kellcuc** [comment]
Verified in amanuensis

**Comment by kellcuc** [comment]
Verified in PcdcAnalysisTools

**Comment by paulmurdoch19** [comment]
@grugna downgrade shouldn't affect anything, the Last PR for this changed it from 4.21 to 4.25 and only changed the pypoetry.toml file

**Comment by kellcuc** [comment]
Verified in gearbox-middleware

**Comment by grugna** [comment]
@kellcuc  The last edge case to test is if guppy is able to validate it since it is using a different library from javascript. Please let me know when the guppy validation is tested in helm.
